aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -0,0 +1,247 @@
+        };
+        borrow_global_mut<ValidatorInfo>(addr).consensus_pubkey = consensus_pubkey;
+    }
+
+    /// Initialize validator set to the core resource account.
+    fun initialize_validator_set(account: &signer, minimum_stake: u64, maximum_stake: u64) {
+        SystemAddresses::assert_core_resource(account);
+        move_to(account, ValidatorSet {
+            consensus_scheme: 0,
+            minimum_stake,
+            maximum_stake,
+            validators: Vector::empty(),
+            last_update_time_secs: Timestamp::now_seconds(),
+        });
+    }
+
+    /// Initiate by the validator info owner
+    fun add_validator(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_info = borrow_global_mut<ValidatorInfo>(addr);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+        assert!(!Vector::contains(&validator_set.validators, &addr), 0);
+
+        assert!(validator_info.stake_info.current_stake >= validator_set.minimum_stake, 0);
+        assert!(validator_info.stake_info.current_stake <= validator_set.maximum_stake, 0);
+
+        on_new_epoch(validator_set);
+        Vector::push_back(&mut validator_set.validators, addr);
+    }
+
+    /// Initiate by the validator info owner.
+    fun remove_validator(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+
+        let (exist, index) = Vector::index_of(&validator_set.validators, &addr);
+        assert!(exist, 0);
+        on_new_epoch(validator_set);
+        Vector::swap_remove(&mut validator_set.validators, index);
+        assert!(Vector::length(&validator_set.validators) > 0, 0);
+    }
+
+    /// Triggers when validator set changes or after certain time period (in block prologue).
+    fun on_new_epoch(validator_set: &mut ValidatorSet) acquires ValidatorInfo {
+        let current_time = Timestamp::now_seconds();
+        assert!(validator_set.last_update_time_secs + MINIMUM_RECONFIG_PERIOD > current_time, 0);
+        validator_set.last_update_time_secs = current_time;
+        let i = 0;
+        let len = Vector::length(&validator_set.validators);
+        while (i < len) {
+            let addr = *Vector::borrow(&validator_set.validators, i);
+            update_stake_info(addr);
+            i = i + 1;
+        };
+        // Reconfiguration::reconfigure();
+        // do we remove validators without enough stakes?
+    }
+
+    /// Update individual validator's stake info
+    /// 1. distribute rewards to active/cooldown delegations
+    /// 2. process warmup, cooldown correspondingly
+    /// 3. update the current stake
+    fun update_stake_info(addr: address) acquires ValidatorInfo {
+        let stake_info = &mut borrow_global_mut<ValidatorInfo>(addr).stake_info;
+        distribute_reward( &mut stake_info.active);
```

Comment:
could you help me understand why having active and cool down? and distribute the reward to both of them?

Reply:
it's probably also bad naming, cool down is pending inactive, someone initiates the withdraw but it doesn't take effect until next epoch.